File: 0410-split-array-largest-sum/0410-split-array-largest-sum.py
class Solution:
    def canSplit(self , nums: List[int] , k : int , maxSum : int) -> bool:

        currentSum = 0
        partitions = 1
        for num in nums:
            if currentSum + num <= maxSum:
                currentSum += num
            else:
                partitions+=1
                currentSum = num
        return partitions <= k                    
    def splitArray(self, nums: List[int], k: int) -> int:

        low = max(nums)
        high = 0
        for num in nums:
            high += num

        ans = 0    

        while low <= high:
            mid = low+(high - low)//2

            if self.canSplit(nums , k , mid):
                ans = mid
                high = mid-1
            else:
                low = mid+1
        return ans               
        

File: 0410-split-array-largest-sum/0410-split-array-largest-sum.java
class Solution {
    private boolean canSplit(int[] nums, int k , int maxSum)
    {
        int currentSum = 0;
        int partitions = 1;

        for(int num : nums)
        {
            if(currentSum + num <= maxSum)
            {
                currentSum += num;
            }
            else
            {
                partitions++;
                currentSum = num;
            }
        }
        return partitions <= k;
    }
    public int splitArray(int[] nums, int k) {

        int low = 0;
        int high = 0;

        for(int num : nums)
        {
            low = Math.max(low , num);
            high += num;
        }
        
        int ans =0;

        while(low <= high)
        {
            int mid = low+(high - low)/2;

            if(canSplit(nums , k , mid))
            {
                ans = mid;
                high = mid-1;
            }
            else
            {
                low = mid+1;
            }
        }
        return ans;
    }
}